generate_workspace/src/main/java/com/google/devtools/build/workspace/output/WorkspaceWriter.java

```diff
@@ -0,0 +1,128 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.workspace.output;
+
+import com.google.devtools.build.workspace.maven.Rule;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.lang.invoke.MethodHandles;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.text.SimpleDateFormat;
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+import java.util.logging.Logger;
+
+/**
+ * Write a set of rules to a WORKSPACE and BUILD file.
+ */
+public class WorkspaceWriter extends AbstractWriter {
+  private final static Logger logger = Logger.getLogger(
+      MethodHandles.lookup().lookupClass().getName());
+
+  private final File workspaceFile;
+  private final File buildFile;
+
+  public WorkspaceWriter(String outputDirStr) throws IOException {
+    Path outputDir;
+    if (outputDirStr.isEmpty()) {
+      outputDir = Files.createTempDirectory("generate_workspace");
+    } else {
+      outputDir = Paths.get(outputDirStr);
+    }
+    String date = new SimpleDateFormat("yyyy-MM-dd-HH-mm-ss").format(new Date());
+    File tempWorkspaceFile = outputDir.resolve("WORKSPACE").toFile();
+    File tempBuildFile = outputDir.resolve("BUILD").toFile();
+
+    // Don't overwrite existing files with generated ones.
+    if (tempWorkspaceFile.exists()) {
+      workspaceFile = outputDir.resolve(date + ".WORKSPACE").toFile();
+    } else {
+      workspaceFile = tempWorkspaceFile;
+    }
+    if (tempBuildFile.exists()) {
+      buildFile = outputDir.resolve(date + ".BUILD").toFile();
+    } else {
+      buildFile = tempBuildFile;
+    }
+  }
+
+  @Override
+  public void write(List<String> sources, Collection<Rule> rules) {
+    try (PrintStream workspaceStream = new PrintStream(workspaceFile);
+        PrintStream buildStream = new PrintStream(buildFile)) {
+      writeWorkspace(workspaceStream, sources, rules);
+      writeBuild(buildStream, sources, rules);
+    } catch (IOException e) {
+      logger.severe(
+          "Could not write WORKSPACE and BUILD files to " + buildFile.getParent() + ": "
+              + e.getMessage());
+      return;
+    }
+    System.out.println("Wrote:\n" + workspaceFile + "\n" + buildFile);
+  }
+
+  /**
+   * Writes all resolved dependencies in WORKSPACE file format to the outputStream.
+   */
+  public void writeWorkspace(PrintStream outputStream, List<String> sources,
+      Collection<Rule> rules) {
+    writeHeader(outputStream, sources);
+    for (Rule rule : rules) {
+      outputStream.println(formatMavenJar(rule) + "\n");
+    }
+  }
+
+  @Override
+  public String formatMavenJar(Rule rule) {
```

Comment:
The logic is complicated enough I think it's still worth unifying, just passing in the initial indent and the rule name.